dspace-server-webapp/src/main/java/org/dspace/app/rest/exception/DSpaceAccessDeniedHandler.java

```diff
@@ -0,0 +1,87 @@
+import org.springframework.security.web.csrf.CsrfToken;
+import org.springframework.security.web.csrf.CsrfTokenRepository;
+import org.springframework.security.web.csrf.InvalidCsrfTokenException;
+import org.springframework.security.web.csrf.MissingCsrfTokenException;
+import org.springframework.stereotype.Component;
+
+/**
+ * This Handler customizes behavior of AccessDeniedException errors thrown by Spring Security/Boot.
+ * <P>
+ * More specifically, we use this Handler to ensure exceptions related to CSRF Tokens are also sent to our
+ * DSpaceApiExceptionControllerAdvice class, which manages all exceptions for the DSpace backend. Without this
+ * handler, those CSRF exceptions are managed by Spring Security/Boot *before* DSpaceApiExceptionControllerAdvice
+ * is triggered.
+ * <P>
+ * Additionally, this Handler is customized to refresh the CSRF Token whenever an InvalidCsrfTokenException occurs.
+ * This helps ensure our DSpace server-side token (stored in a server-side cookie) remains "synced" with the token
+ * on the client side. If they ever get out of sync, the next request will throw an InvalidCsrfTokenException.
+ *
+ * @see DSpaceApiExceptionControllerAdvice
+ */
+@Component
+public class DSpaceAccessDeniedHandler implements AccessDeniedHandler {
+
+    @Autowired
+    private WebSecurityConfiguration webSecurityConfiguration;
+
+    /**
+     * Override handle() to pass these exceptions over to our DSpaceApiExceptionControllerAdvice handler
+     * @param request request
+     * @param response response
+     * @param ex AccessDeniedException
+     * @throws IOException
+     * @throws ServletException
+     */
+    @Override
+    public void handle(HttpServletRequest request, HttpServletResponse response, AccessDeniedException ex)
+        throws IOException, ServletException {
+
+        // Do nothing if response is already committed
+        if (response.isCommitted()) {
+            return;
+        }
+
+        // If we had an InvalidCsrfTokenException, this means the client sent a CSRF token which did *not* match the
+        // token on the server. In this scenario, we trigger a refresh of the CSRF token...as it's possible the user
+        // switched clients (from HAL Browser to UI or visa versa) and has an out-of-sync token.
+        if (ex instanceof InvalidCsrfTokenException) {
+            // Get access to our enabled CSRF token repository
+            CsrfTokenRepository csrfTokenRepository = webSecurityConfiguration.getCsrfTokenRepository();
+
+            // Remove current token & generate a new one
+            csrfTokenRepository.saveToken(null, request, response);
+            CsrfToken newToken = csrfTokenRepository.generateToken(request);
+            csrfTokenRepository.saveToken(newToken, request, response);
+        }
+
+        // Get access to our general exception handler
+        DSpaceApiExceptionControllerAdvice handler = new DSpaceApiExceptionControllerAdvice();
```

Comment:
just a minor thing but I would be happier if we found a way to resolve it at runtime or inject it. We should have a way to get the ExceptionHandlerResolver from spring and pass to it the exception

Reply:
@abollini : I'm not sure this is very easy.  Spring Security is what manages AccessDeniedExceptions, and it catches them & responds *before* anything gets to Spring MVC.  I've yet to figure out a way to do this at runtime simply because the exception is handled so early in the request.  

So, your question is similar to the one answered here: https://stackoverflow.com/questions/41140669/handle-security-exceptions-in-spring-boot-resource-server  (and the first response is that you cannot manage this sort of exception in the MVC layer.

If you see another option here, I'm willing to try something else.  I just couldn't figure out an easy way around just calling `DSpaceApiExceptionControllerAdvice()` directly. The only other option I thought of is to remove that call & move the exceptions into `DSpaceAccessDeniedHandler` (but then we'd have some exceptions handled in `DSpaceApiExceptionControllerAdvice` and others in `DSpaceAccessDeniedHandler`)

Reply:
@abollini : Please disregard my previous response. Turns out, there is a way to do this more dynamically.  I stumbled on an idea [here](https://stackoverflow.com/questions/41140669/handle-security-exceptions-in-spring-boot-resource-server#comment98388793_41303283) that led me down the right path...and it seems to be exactly what you were implying.  So, see the latest commit: https://github.com/DSpace/DSpace/pull/3103/commits/70253bc1bf16ce6b612e5bbf3cc876407e05636a  (includes a small test to ensure our custom error message is coming back -- I also tested it manually and it works)